Tests/GRPCHTTP2CoreTests/Client/Resolver/NameResolverRegistryTests.swift

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import GRPCCore
+import GRPCHTTP2Core
+import XCTest
+
+final class NameResolverRegistryTests: XCTestCase {
+  struct FailingResolver: NameResolverFactory {
+    typealias Target = StringTarget
+
+    func resolver(for target: NameResolverRegistryTests.StringTarget) -> NameResolver {
+      let stream = AsyncThrowingStream(NameResolutionResult.self) {
+        $0.yield(with: .failure(RPCError(code: .unavailable, message: target.value)))
+      }
+
+      return NameResolver(results: RPCAsyncSequence(wrapping: stream), updateMode: .pull)
+    }
+  }
+
+  struct StringTarget: ResolvableTarget {
+    var value: String
+
+    init(value: String) {
+      self.value = value
+    }
+  }
+
+  func testEmptyNameResolvers() {
+    let resolvers = NameResolverRegistry()
+    XCTAssert(resolvers.isEmpty)
+    XCTAssertEqual(resolvers.count, 0)
+  }
+
+  func testRegisterFactory() {
+    var resolvers = NameResolverRegistry()
+    resolvers.registerFactory(FailingResolver())
+    XCTAssertEqual(resolvers.count, 1)
+
+    // Adding a resolver of the same type replaces it.
+    resolvers.registerFactory(FailingResolver())
```

Comment:
May be a bit overkill, but we're not exactly testing that the new resolver replaces the old one here: we're only asserting that there's just one left. You think we can initialise the `FailingResolver` with some form of ID, so that we can assert the new one is the one that remains?